src/Orleans.Persistence.CosmosDB/Options/CosmosDBStorageOptions.cs

```diff
@@ -59,6 +63,15 @@ public class CosmosDBStorageOptions
         /// Stage of silo lifecycle where storage should be initialized.  Storage must be initialized prior to use.
         /// </summary>
         public int InitStage { get; set; } = DEFAULT_INIT_STAGE;
+
+        /// <summary>
+        /// Configurable provider for partition key usage 
+        /// </summary>
+        /// <param name="type">Graintype</param>
+        /// <param name="partitionKeyBuilder">Function for creating the partition key based on the GrainReference</param>
+        /// <returns>String used as partition key</returns>
+        public IPartitionKeyProvider PartitionKeyProvider { get; set; } = new DefaultPartitionKeyProvider();

```

Comment:
Now that it is an interface, this must be removed from the Options class and instead registered on DI.

The `DefaultPartitionKeyProvider` should be registered thru DI using `TryAddSingleton<IPartitionKeyProvider, DefaultPartitionKeyProvider>`. That ensure that if an user has not registered one, the default implementation is used. 

Also make sure you inject it on the provider's ctor.

Reply:
Do we really want this in DI?

It is something I would like very much in the clear and open. The default instance is set on the CosmosDBOptions, if one would like to override it setting it explicitly on the options makes it very clear that it is in fact overridden and with what. 

I have seen too many DI's gone wrong due to duplicate registration, assembly scanning and obfuscation.

Reply:
Yes, that is the whole point of having the interface.

The options types are created for config-only. They can be serialized and configured (for example) on appsettings.json. How would you tell from JSON that you want to use implementation A or B? It is an implementation detail, not a configuration as is.

Regarding the DI's gone wrong, probably is because it wasn't implemented correctly.

Orleans is fully DI-based on its dependencies and the same happens on ASP.Net Core and the whole .Net Core ecosystem. There is no assembly scanning or obfuscation. DI doesn't do that.

It is ok if you don't want to add it, I can change it myself in another PR, just let me know and I can merge it as is.